Drop StringView::characters8() / characters16()

cdumez opened the pull request:
#### b332f986be2c69a696e8ab2646b087123fc64cc4
<pre>
Drop StringView::characters8() / characters16()
<a href="https://bugs.webkit.org/show_bug.cgi?id=273052">https://bugs.webkit.org/show_bug.cgi?id=273052</a>

Reviewed by Darin Adler.

Drop StringView::characters8() / characters16(), in favor to span8() / span16().

* Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
(JSC::BuiltinExecutables::createExecutable):
* Source/JavaScriptCore/parser/Lexer.h:
(JSC::Lexer&lt;LChar&gt;::setCodeStart):
(JSC::Lexer&lt;UChar&gt;::setCodeStart):
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::fastJoin):
* Source/JavaScriptCore/runtime/IntlCollator.cpp:
(JSC::IntlCollator::compareStrings const):
(JSC::IntlCollator::checkICULocaleInvariants):
* Source/JavaScriptCore/runtime/IntlObjectInlines.h:
(JSC::followedByNonLatinCharacter):
(JSC::compareASCIIWithUCADUCETLevel3):
(JSC::compareASCIIWithUCADUCET):
* Source/JavaScriptCore/runtime/JSBigInt.cpp:
(JSC::JSBigInt::parseInt):
* Source/JavaScriptCore/runtime/JSBigInt.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSImmutableButterfly.cpp:
(JSC::JSImmutableButterfly::createFromString):
* Source/JavaScriptCore/runtime/ParseInt.h:
(JSC::parseInt):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
(JSC::Yarr::SubjectSampler::sample):
* Source/WTF/wtf/text/StringSearch.h:
(WTF::BoyerMooreHorspoolTable::BoyerMooreHorspoolTable):
* Source/WTF/wtf/text/StringView.h:
* Source/WebCore/css/CSSFontFaceSet.cpp:
(WebCore::codePointsFromString):
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::checkForValidDouble):
(WebCore::parseDouble):
(WebCore::parseColorIntOrPercentage):
(WebCore::isTenthAlpha):
(WebCore::parseAlphaValue):
(WebCore::mightBeRGBA):
(WebCore::mightBeRGB):
(WebCore::parseHexColorInternal):
(WebCore::parseNumericColor):
(WebCore::parseNamedColorInternal):
(WebCore::parseSimpleColorInternal):
(WebCore::CSSParserFastPaths::parseSimpleColor):
(WebCore::CSSParserFastPaths::parseHexColor):
(WebCore::CSSParserFastPaths::parseNamedColor):
* Source/WebCore/css/parser/CSSParserToken.cpp:
(WebCore::mergeIfAdjacent):
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::attachmentTitleForDisplay const):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::enclosingGlyphBoundsForText):
(WebCore::Layout::TextUtil::containsStrongDirectionalityText):
(WebCore::Layout::TextUtil::directionForTextContent):
(WebCore::Layout::TextUtil::canUseSimplifiedTextMeasuring):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::widthForSimpleTextSlow const):
(WebCore::FontCascade::codePath const):
(WebCore::FontCascade::characterRangeCodePath):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/TextRun.h:
* Source/WebCore/platform/network/create-http-header-name-table:
* Source/WebCore/platform/sql/SQLiteStatement.cpp:
(WebCore::SQLiteStatement::bindText):
* Source/WebCore/platform/text/SegmentedString.h:
(WebCore::SegmentedString::Substring::Substring):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::computeCanUseSimpleFontCodePath const):
* Source/WebKit/Shared/API/c/WKString.cpp:
(WKStringGetUTF8CStringImpl):
* Source/WebKit/WebProcess/GPU/media/RemoteVideoCodecFactory.cpp:
(WebKit::RemoteVideoCodecFactory::createEncoder):
* Source/WebKit/WebProcess/InjectedBundle/API/mac/WKDOMTextIterator.mm:
(-[WKDOMTextIterator currentTextPointer]):
* Source/WebKitLegacy/mac/WebView/WebTextIterator.mm:
(-[WebTextIterator currentTextPointer]):
* Tools/TestWebKitAPI/Tests/WTF/StringParsingBuffer.cpp:
(TestWebKitAPI::TEST(WTF, StringParsingBufferInitial)):

Canonical link: <a href="https://commits.webkit.org/277836@main">https://commits.webkit.org/277836@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0efc559e5c20bd73d5514114655123b6f43d2978

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48704 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27915 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51665 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51391 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44769 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51009 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33852 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25445 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39832 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/49286 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25570 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42026 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20929 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23045 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/43203 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6760 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/42003 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44986 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/43693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/53300 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/48195 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23751 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20045 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47126 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25017 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42231 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46058 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25821 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/55690 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6951 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24738 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11458 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
```diff
@@ -605,24 +605,20 @@ JSC_DEFINE_HOST_FUNCTION(globalFuncEscape, (JSGlobalObject* globalObject, CallFr
-                    builder.append(*c);
+            for (auto character : view.span16()) {
+                if (character >= doNotEscape.size())
+                    builder.append("%u", hex(static_cast<uint8_t>(character >> 8), 2), hex(static_cast<uint8_t>(character), 2));
```
`"%u"_s`

darinadler commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -323,72 +325,72 @@ static std::optional<uint8_t> parseColorIntOrPercentage(const CharacterType*& st
-            string = end;
+    if (string.front() != '0' && string.front() != '1' && string.front() != '.') {
+        if (checkForValidDouble(string, terminator)) {
+            string = string.subspan(string.size());
```
We could just set it to `{ }` perhaps?

cdumez replied:
Will give it a try.

darinadler commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -323,72 +325,72 @@ static std::optional<uint8_t> parseColorIntOrPercentage(const CharacterType*& st
-        string = end;
+    if (length == 2 && string.front() != '.') {
+        uint8_t result = !negative && string.front() == '1' ? 255 : 0;
+        string = string.subspan(string.size());
```
We could just set it to `{ }` perhaps?

darinadler commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -323,72 +325,72 @@ static std::optional<uint8_t> parseColorIntOrPercentage(const CharacterType*& st
         static constexpr uint8_t tenthAlphaValues[] = { 0, 26, 51, 77, 102, 128, 153, 179, 204, 230 };
         uint8_t result = negative ? 0 : tenthAlphaValues[string[length - 2] - '0'];
-        string = end;
+        string = string.subspan(string.size());
```
We could just set it to `{ }` perhaps?

darinadler commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -323,72 +325,72 @@ static std::optional<uint8_t> parseColorIntOrPercentage(const CharacterType*& st
         return std::nullopt;
 
-    string = end;
+    string = string.subspan(string.size());
```
We could just set it to `{ }` perhaps?

darinadler commented on Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
```diff
@@ -220,10 +220,10 @@ TextUtil::EnclosingAscentDescent TextUtil::enclosingGlyphBoundsForText(StringVie
 
     if (textContent.is8Bit()) {
-        auto textIterator = Latin1TextIterator { textContent.characters8(), 0, textContent.length(), textContent.length() };
+        auto textIterator = Latin1TextIterator { textContent.span8().data(), 0, textContent.length(), textContent.length() };
```
Seems like these iterators should be taking spans; I know we say this all the time when reviewing these.

cdumez replied:
Yes, I opted out of doing it in this patch because it was already large though.

darinadler commented on Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
```diff
@@ -220,10 +220,10 @@ TextUtil::EnclosingAscentDescent TextUtil::enclosingGlyphBoundsForText(StringVie
         return enclosingGlyphBoundsForRunWithIterator(style.fontCascade(), !style.isLeftToRightDirection(), textIterator);
     }
-    auto textIterator = SurrogatePairAwareTextIterator { textContent.characters16(), 0, textContent.length(), textContent.length() };
+    auto textIterator = SurrogatePairAwareTextIterator { textContent.span16().data(), 0, textContent.length(), textContent.length() };
```
Seems like these iterators should be taking spans; I know we say this all the time when reviewing these.

darinadler commented on Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
```diff
@@ -460,7 +460,8 @@ bool TextUtil::containsStrongDirectionalityText(StringView text)
         char32_t character;
-        U16_NEXT(text.characters16(), position, length, character);
+        auto span = text.span16();
+        U16_NEXT(span, position, length, character);
```
I assume that `span16()` is a simple enough function, and inlined, so we can just pass `text.span16()` to `U16_NEXT` instead of using a local variable.

cdumez replied:
I'll try again. I seem to remember that the last time I did that I got some warning about some unused `__c2` variable on the bots and not locally though.

darinadler replied:
Apple has modified our ICU to resolve an unused `__c2` warning, you can see that looking in the Apple copy of the ICU `utf16.h` header.

cdumez replied:
Ok, I'll give it another try. I saw the error on the bots last week though.

darinadler replied:
Maybe bots from non-Apple platforms. This is not important; don’t waste too much time on it please?

cdumez replied:
Still failing on the Mac bots: https://ews-build.webkit.org/#/builders/51/builds/25437

darinadler commented on Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
```diff
@@ -499,7 +500,7 @@ TextDirection TextUtil::directionForTextContent(StringView content)
     if (content.is8Bit())
         return TextDirection::LTR;
-    return ubidi_getBaseDirection(content.characters16(), content.length()) == UBIDI_RTL ? TextDirection::RTL : TextDirection::LTR;
+    return ubidi_getBaseDirection(content.span16().data(), content.length()) == UBIDI_RTL ? TextDirection::RTL : TextDirection::LTR;
```
Seems like the many places like this should use arguments like `content.span16().size()` instead of `content.length()`.

darinadler commented on Source/WebCore/platform/graphics/FontCascade.h:
```diff
@@ -217,8 +217,8 @@ class FontCascade final : public CanMakeWeakPtr<FontCascade>, public CanMakeChec
     WEBCORE_EXPORT CodePath codePath(const TextRun&, std::optional<unsigned> from = std::nullopt, std::optional<unsigned> to = std::nullopt) const;
-    static CodePath characterRangeCodePath(const LChar*, unsigned) { return CodePath::Simple; }
-    static CodePath characterRangeCodePath(const UChar*, unsigned len);
+    static CodePath characterRangeCodePath(std::span<const LChar>) { return CodePath::Simple; }
```
I almost want to put `constexpr` on this line, but it probably has no benefit.

darinadler commented on Source/WebCore/platform/network/create-http-header-name-table:
```diff
@@ -137,21 +137,20 @@ bool findHTTPHeaderName(StringView stringView, HTTPHeaderName& headerName)
-        LChar characters[maxHTTPHeaderNameLength];
-        for (unsigned i = 0; i < length; ++i) {
-            UChar character = stringView.characters16()[i];
+        char characters[maxHTTPHeaderNameLength];
```
Should this use `std::array`? Probably not that beneficial since it won’t know the length at compile time. Would be nice if we had a `std::span` with the correct length instead of passing a pointer/length pair. Almost want to just make a `std::array` and a `std::span` of the correct length up front, but that’s not really all that great.

darinadler commented on Source/WebCore/platform/network/create-http-header-name-table:
```diff
@@ -137,21 +137,20 @@ bool findHTTPHeaderName(StringView stringView, HTTPHeaderName& headerName)
+        char characters[maxHTTPHeaderNameLength];
+        size_t index = 0;
+        for (auto character : stringView.span16()) {
             if (!isASCII(character))
```
Seems like we could do the `containsOnlyASCII` check separately and then not have to write a loop just to put the characters into an array. This can’t be the only place that wants to get a `StringView` into a character array; I suppose it’s optimized to only check for ASCII in the 16-bit case.

darinadler commented on Source/WebCore/platform/sql/SQLiteStatement.cpp:
```diff
@@ -123,9 +123,10 @@ int SQLiteStatement::bindText(int index, StringView text)
-
+    if (text.is8Bit() && text.containsOnlyASCII()) {
+        auto characters = spanReinterpretCast<const char>(text.span8());
+        return sqlite3_bind_text(m_statement, index, characters.size() ? characters.data() : "", characters.size(), SQLITE_TRANSIENT);
```
Maybe `characters.empty() ? "" : characters.data()`

darinadler commented on Source/WebKit/WebProcess/GPU/media/RemoteVideoCodecFactory.cpp:
```diff
@@ -177,8 +177,8 @@ void RemoteVideoCodecFactory::createEncoder(const String& codec, const WebCore::
         if (auto position = codec.find('.');position != notFound && position != codec.length()) {
-            auto profileLevelId = StringView(codec).substring(position + 1);
-            parameters["profile-level-id"] = std::string(reinterpret_cast<const char*>(profileLevelId.characters8()), profileLevelId.length());
+            auto profileLevelId = spanReinterpretCast<const char>(StringView(codec).substring(position + 1).span8());
```
I don’t think `StringView` needs to be involved here. Instead `codec.span8().subspan(position + 1)`

cdumez committed:
Drop StringView::characters8() / characters16()
https://bugs.webkit.org/show_bug.cgi?id=273052

Reviewed by Darin Adler.

Drop StringView::characters8() / characters16(), in favor to span8() / span16().

* Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
(JSC::BuiltinExecutables::createExecutable):
* Source/JavaScriptCore/parser/Lexer.h:
(JSC::Lexer<LChar>::setCodeStart):
(JSC::Lexer<UChar>::setCodeStart):
* Source/JavaScriptCore/runtime/ArrayPrototype.cpp:
(JSC::fastJoin):
* Source/JavaScriptCore/runtime/IntlCollator.cpp:
(JSC::IntlCollator::compareStrings const):
(JSC::IntlCollator::checkICULocaleInvariants):
* Source/JavaScriptCore/runtime/IntlObjectInlines.h:
(JSC::followedByNonLatinCharacter):
(JSC::compareASCIIWithUCADUCETLevel3):
(JSC::compareASCIIWithUCADUCET):
* Source/JavaScriptCore/runtime/JSBigInt.cpp:
(JSC::JSBigInt::parseInt):
* Source/JavaScriptCore/runtime/JSBigInt.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectFunctions.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/JSImmutableButterfly.cpp:
(JSC::JSImmutableButterfly::createFromString):
* Source/JavaScriptCore/runtime/ParseInt.h:
(JSC::parseInt):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
(JSC::Yarr::SubjectSampler::sample):
* Source/WTF/wtf/text/StringSearch.h:
(WTF::BoyerMooreHorspoolTable::BoyerMooreHorspoolTable):
* Source/WTF/wtf/text/StringView.h:
* Source/WebCore/css/CSSFontFaceSet.cpp:
(WebCore::codePointsFromString):
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::checkForValidDouble):
(WebCore::parseDouble):
(WebCore::parseColorIntOrPercentage):
(WebCore::isTenthAlpha):
(WebCore::parseAlphaValue):
(WebCore::mightBeRGBA):
(WebCore::mightBeRGB):
(WebCore::parseHexColorInternal):
(WebCore::parseNumericColor):
(WebCore::parseNamedColorInternal):
(WebCore::parseSimpleColorInternal):
(WebCore::CSSParserFastPaths::parseSimpleColor):
(WebCore::CSSParserFastPaths::parseHexColor):
(WebCore::CSSParserFastPaths::parseNamedColor):
* Source/WebCore/css/parser/CSSParserToken.cpp:
(WebCore::mergeIfAdjacent):
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::attachmentTitleForDisplay const):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::enclosingGlyphBoundsForText):
(WebCore::Layout::TextUtil::containsStrongDirectionalityText):
(WebCore::Layout::TextUtil::directionForTextContent):
(WebCore::Layout::TextUtil::canUseSimplifiedTextMeasuring):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::widthForSimpleTextSlow const):
(WebCore::FontCascade::codePath const):
(WebCore::FontCascade::characterRangeCodePath):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/TextRun.h:
* Source/WebCore/platform/network/create-http-header-name-table:
* Source/WebCore/platform/sql/SQLiteStatement.cpp:
(WebCore::SQLiteStatement::bindText):
* Source/WebCore/platform/text/SegmentedString.h:
(WebCore::SegmentedString::Substring::Substring):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::computeCanUseSimpleFontCodePath const):
* Source/WebKit/Shared/API/c/WKString.cpp:
(WKStringGetUTF8CStringImpl):
* Source/WebKit/WebProcess/GPU/media/RemoteVideoCodecFactory.cpp:
(WebKit::RemoteVideoCodecFactory::createEncoder):
* Source/WebKit/WebProcess/InjectedBundle/API/mac/WKDOMTextIterator.mm:
(-[WKDOMTextIterator currentTextPointer]):
* Source/WebKitLegacy/mac/WebView/WebTextIterator.mm:
(-[WebTextIterator currentTextPointer]):
* Tools/TestWebKitAPI/Tests/WTF/StringParsingBuffer.cpp:
(TestWebKitAPI::TEST(WTF, StringParsingBufferInitial)):

Canonical link: https://commits.webkit.org/277836@main
